pytorch_lightning/loops/utilities.py

```diff
@@ -40,76 +39,51 @@ def check_finite_loss(loss: Optional[torch.Tensor]) -> None:
+            "a dict with key 'loss' or None (where the step will be skipped)."
+        )
+
+
+def _extract_hiddens(training_step_output: STEP_OUTPUT, truncated_bptt_steps: int) -> Optional[Any]:
+    """Get the hidden embeddings if present from the training step output.
+
+    Raises:
+        MisconfigurationException: If :attr:`~pytorch_lightning.core.Lightning.LightningModule.truncated_bptt_steps` is
+            not enabled and hiddens are returned or vice versa.
+    """
+    if not truncated_bptt_steps:
+        if "hiddens" in training_step_output:
             raise MisconfigurationException(
-                "In automatic optimization, `training_step` must either return a Tensor, "
-                "a dict with key 'loss' or None (where the step will be skipped)."
+                'You returned "hiddens" in your `training_step` but `truncated_bptt_steps` is disabled'
             )
-
-
-def _process_training_step_output(
-    trainer: "pl.Trainer", training_step_output: STEP_OUTPUT
-) -> Tuple[Optional[ResultCollection], Optional[Any]]:
-    """Adds the :param:`training_step_output` to the trainer's results.
-
-    Args:
-        trainer: a reference to the trainer
-        training_step_output: the output of the training step (before wrapping into an AttributeDict)
-
-    Returns:
-        the updated results (None if the training_step's output was None) and hiddens exract from the results
-    """
-    if training_step_output is None:
-        return None, None
-
-    results = trainer._results
-
-    loss = None
-    hiddens = None
-
-    # handle dict return
-    if isinstance(training_step_output, dict):
-        # this should not modify the `training_step_output`, as the user could be using it after `training_step_end`
-        loss = training_step_output.get("loss")
-        hiddens = training_step_output.get("hiddens")
-        # detach hiddens to avoid `RuntimeError: Trying to backward through the graph a second time`
-        hiddens = apply_to_collection(hiddens, torch.Tensor, lambda t: t.detach())
-        # use the setter instead of `dict.update` because it calls `detach` on the tensor items
-        results.extra = {k: v for k, v in training_step_output.items() if k not in ("loss", "hiddens")}
-
-    # handle scalar return
-    elif isinstance(training_step_output, torch.Tensor):
-        loss = training_step_output
-
-    if trainer.terminate_on_nan:
-        check_finite_loss(loss)
-
-    # the loss shouldn't be moved to cpu.
-    if trainer.move_metrics_to_cpu:
-        results.cpu()
-
-    # map to results under the hood
-    results.minimize = loss
-
-    return results, hiddens
+        return
+    elif not isinstance(training_step_output, dict) or "hiddens" not in training_step_output:
+        raise MisconfigurationException(
+            'You enabled `truncated_bptt_steps` but did not return "hiddens" in your `training_step`'
+        )
```

Comment:
These docs

https://github.com/PyTorchLightning/pytorch-lightning/blob/439db224d8d645d0f7c962e71a2e46b176c652c7/pytorch_lightning/core/lightning.py#L249

say this is expected, so I'm enforcing it by raising an error otherwise